pandas/core/reshape/reshape.py

```diff
@@ -1096,6 +1100,162 @@ def get_empty_frame(data) -> DataFrame:
+        raise ValueError(
+            f"Dummy DataFrame contains NA value in column: "  # type: ignore[union-attr]
+            f"'{data.isna().any().idxmax()}'"
+        )
+
+    # index data with a list of all columns that are dummies
+    try:
+        data_to_decode = data.astype("boolean")
+    except TypeError:
+        raise TypeError("Passed DataFrame contains non-dummy data")
+
+    # collect prefixes and get lists to slice data for each prefix
+    variables_slice = defaultdict(list)
+    if sep is None:
+        variables_slice[""] = list(data.columns)
+    elif isinstance(sep, str):
+        for col in data_to_decode.columns:
+            prefix = col.split(sep)[0]
+            if len(prefix) == len(col):
+                raise ValueError(f"Separator not specified for column: {col}")
+            variables_slice[prefix].append(col)
+    else:
+        raise TypeError(
+            f"Expected 'sep' to be of type 'str' or 'None'; "
+            f"Received 'sep' of type: {type(sep).__name__}"
+        )
+
+    # validate number of implied_category
+    def check_len(item, name) -> None:
+        if not len(item) == len(variables_slice):
+            len_msg = (
+                f"Length of '{name}' ({len(item)}) did not match the "
+                "length of the columns being encoded "
+                f"({len(variables_slice)})"
+            )
+            raise ValueError(len_msg)
+
+    if implied_category:
+        if isinstance(implied_category, dict):
+            check_len(implied_category, "implied_category")
+        elif isinstance(implied_category, Hashable):
+            implied_category = dict(
+                zip(variables_slice, [implied_category] * len(variables_slice))
+            )
+        else:
+            raise TypeError(
+                f"Expected 'implied_category' to be of type "
+                f"'None', 'Hashable', or 'dict'; "
+                f"Received 'implied_category' of type: "
+                f"{type(implied_category).__name__}"
+            )
+
+    cat_data = {}
+    for prefix, prefix_slice in variables_slice.items():
+        if sep is None:
+            cats = prefix_slice.copy()
+        else:
+            cats = [col[len(prefix + sep) :] for col in prefix_slice]
+        assigned = data_to_decode[prefix_slice].sum(axis=1)
+        if any(assigned > 1):
+            raise ValueError(
+                f"Dummy DataFrame contains multi-assignment(s); "
+                f"First instance in row: {assigned.idxmax()}"
+            )
+        elif any(assigned == 0):
+            if isinstance(implied_category, dict):
+                cats.append(implied_category[prefix])
+            else:
+                raise ValueError(
+                    f"Dummy DataFrame contains unassigned value(s); "
+                    f"First instance in row: {assigned.idxmin()}"
+                )
+            data_slice = concat((data_to_decode[prefix_slice], assigned == 0), axis=1)
```

Comment:
maybe i am missing something, but this loop can overwrite `data_slice`? that seems odd

Reply:
The `data_slice` changes during each iteration as it contains the category indices of each element for each category group (as grouped by the prefixes). After each group is added to the decoded categorical data, it is no longer required and `data_slice` can be reused.